File: adapter/harness/turn_context.py
"""
G-2 — TurnContextBootstrap

Configurable turn-state initialiser for multi-turn conversational agents.
Handles source-path field mapping, turn-1-only seeding, resource budget
tracking, and model skeleton initialisation.
"""

from __future__ import annotations

import copy
from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any


@dataclass
class SessionField:
    key: str
    default: Any
    source_path: str | None = None
    init_once: bool = False


@dataclass
class ResourceBudget:
    time_limit_seconds: int
    token_budget: int
    budget_key: str = "resource_budget"
    turn_key: str = "turn_number"


def _resolve_dot_path(state: dict, path: str) -> tuple[bool, Any]:
    """Traverse a dot-separated path in state. Returns (found, value)."""
    current = state
    for part in path.split("."):
        if not isinstance(current, dict) or part not in current:
            return False, None
        current = current[part]
        if current is None:
            return True, None
    return True, current


def make_turn_initializer(
    fields: list[SessionField],
    resource_budget: ResourceBudget | None = None,
    empty_model_key: str | None = None,
    empty_model_template: dict | None = None,
) -> Callable[[dict], dict]:
    turn_key = resource_budget.turn_key if resource_budget is not None else "turn_number"

    def initializer(state: dict) -> dict:
        result = {**state}

        raw_turn = result.get(turn_key)
        is_turn_one = raw_turn is None or raw_turn <= 1

        # Steps 1 & 2: SessionFields
        for f in fields:
            if f.source_path is not None:
                if f.init_once and not is_turn_one:
                    continue
                found, value = _resolve_dot_path(result, f.source_path)
                if not found or value is None:
                    value = f.default
                result[f.key] = value
            else:
                # Write default only if key is absent (init_once has no effect here)
                if f.key not in result:
                    result[f.key] = f.default

        # Step 3: Resource budget
        if resource_budget is not None:
            budget_key = resource_budget.budget_key
            now = datetime.now(UTC)

            if is_turn_one:
                result[budget_key] = {
                    "time_limit_seconds": resource_budget.time_limit_seconds,
                    "token_budget": resource_budget.token_budget,
                    "elapsed_seconds": 0,
                    "tokens_used": 0,
                    "started_at": now,
                }
            else:
                existing = result.get(budget_key, {})
                started_at = existing.get("started_at")
                elapsed = (now - started_at).total_seconds() if started_at is not None else 0
                result[budget_key] = {**existing, "elapsed_seconds": elapsed}

        # Step 4: Empty model seeding
        if empty_model_key is not None and not result.get(empty_model_key):
            template = empty_model_template if empty_model_template is not None else {}
            result[empty_model_key] = copy.deepcopy(template)

        return result

    return initializer

File: packages/harness/src/primitives/turn-context.ts
/**
 * G-2 — TurnContextBootstrap
 *
 * Configurable turn-state initialiser for multi-turn conversational agents.
 * Port of adapter/harness/turn_context.py. No domain vocabulary.
 */

export interface SessionField {
  key: string
  default: unknown
  sourcePath?: string
  initOnce?: boolean
}

export interface ResourceBudget {
  timeLimitSeconds: number
  tokenBudget: number
  budgetKey?: string
  turnKey?: string
}

function resolveDotPath(
  state: Record<string, unknown>,
  path: string,
): { found: boolean; value: unknown } {
  let current: unknown = state
  for (const part of path.split('.')) {
    if (typeof current !== 'object' || current === null || !(part in (current as Record<string, unknown>))) {
      return { found: false, value: undefined }
    }
    current = (current as Record<string, unknown>)[part]
    if (current === null || current === undefined) {
      return { found: true, value: null }
    }
  }
  return { found: true, value: current }
}

export function makeTurnInitializer(
  fields: SessionField[],
  options?: {
    resourceBudget?: ResourceBudget
    emptyModelKey?: string
    emptyModelTemplate?: Record<string, unknown>
  },
): (state: Record<string, unknown>) => Record<string, unknown> {
  const turnKey = options?.resourceBudget?.turnKey ?? 'turn_number'

  return function initializer(state: Record<string, unknown>): Record<string, unknown> {
    const result: Record<string, unknown> = { ...state }
    const rawTurn = result[turnKey] as number | null | undefined
    const isTurnOne = rawTurn == null || rawTurn <= 1

    // Steps 1 & 2: SessionFields
    for (const f of fields) {
      if (f.sourcePath !== undefined) {
        if (f.initOnce && !isTurnOne) continue
        const { found, value } = resolveDotPath(result, f.sourcePath)
        result[f.key] = !found || value === null || value === undefined ? f.default : value
      } else {
        if (!(f.key in result)) {
          result[f.key] = f.default
        }
      }
    }

    // Step 3: Resource budget
    if (options?.resourceBudget !== undefined) {
      const budget = options.resourceBudget
      const budgetKey = budget.budgetKey ?? 'resource_budget'

      if (isTurnOne) {
        result[budgetKey] = {
          timeLimitSeconds: budget.timeLimitSeconds,
          tokenBudget: budget.tokenBudget,
          elapsedSeconds: 0,
          tokensUsed: 0,
          startedAt: new Date(),
        }
      } else {
        const existing = (result[budgetKey] ?? {}) as Record<string, unknown>
        const startedAt = existing['startedAt']
        const elapsed = startedAt instanceof Date ? (Date.now() - startedAt.getTime()) / 1000 : 0
        result[budgetKey] = { ...existing, elapsedSeconds: elapsed }
      }
    }

    // Step 4: Empty model seeding
    if (options?.emptyModelKey !== undefined && !result[options.emptyModelKey]) {
      const template = options.emptyModelTemplate ?? {}
      result[options.emptyModelKey] = JSON.parse(JSON.stringify(template))
    }

    return result
  }
}
